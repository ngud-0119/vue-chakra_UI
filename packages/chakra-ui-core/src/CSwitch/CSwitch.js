/**
 * Hey! Welcome to @chakra-ui/vue CSwitch
 *
 * The Switch component is used as an alternative for the Checkbox
 * component, switch between enabled or disabled states.
 *
 * @see Docs     https://vue.chakra-ui.com/stat
 * @see Source   https://github.com/chakra-ui/chakra-ui-vue/blob/master/packages/chakra-ui-core/src/CSwitch/CSwitch.js
 */

import styleProps from '../config/props'
import { forwardProps } from '../utils'

import CBox from '../CBox'
import CVisuallyHidden from '../CVisuallyHidden'
import CControlBox from '../CControlBox'

const switchSizes = {
  sm: {
    width: '1.375rem',
    height: '0.75rem'
  },
  md: {
    width: '1.875rem',
    height: '1rem'
  },
  lg: {
    width: '2.875rem',
    height: '1.5rem'
  }
}

/**
 * CSwitch component
 *
 * the switch component that serves as an alternative for checkbox.
 *
 * @extends CBox
 * @see Docs https://vue.chakra-ui.com/stat
 */
const CSwitch = {
  name: 'CSwitch',
  model: {
    prop: 'isChecked',
    event: 'change'
  },
  inject: ['$chakraColorMode'],
  props: {
    ...styleProps,
    id: String,
    name: String,
    value: Boolean,
    _ariaLabel: String,
    _ariaLabelledBy: String,
    color: {
      type: String,
      default: 'blue'
    },
    defaultIsChecked: Boolean,
    isChecked: Boolean,
    size: {
      type: String,
      default: 'md'
    },
    isDisabled: Boolean,
    isInvalid: Boolean
  },
  computed: {
    colorMode () {
      return this.$chakraColorMode()
    },
    _width () {
      return switchSizes[this.size] && switchSizes[this.size].width
    },
    _height () {
      return switchSizes[this.size] && switchSizes[this.size].height
    },
    styleProps () {
      return {
        rounded: 'full',
        justifyContent: 'flex-start',
        width: this._width,
        height: this._height,
        bg: this.colorMode === 'dark' ? 'whiteAlpha.400' : 'gray.300',
        boxSizing: 'content-box',
        p: '2px',
        _checked: {
          bg: `${this.color}.500`
        },
        _child: {
          transform: 'translateX(0)'
        },
        _checkedAndChild: {
          transform: `translateX(calc(${this._width} - ${this._height}))`
        },
        _focus: {
          boxShadow: 'outline'
        },
        _hover: {
          cursor: 'pointer'
        },
        _checkedAndHover: {
          cursor: 'pointer'
        },
        _disabled: {
          opacity: 0.4,
          cursor: 'not-allowed'
        }
      }
    }
  },
  render (h) {
    return h(CBox, {
      props: {
        ...forwardProps(this.$props),
        as: 'label',
        display: 'inline-block',
        verticalAlign: 'middle'
      },
      attrs: {
        'data-chakra-component': 'CSwitch'
      }
    }, [
      h(CVisuallyHidden, {
        props: {
          as: 'input'
        },
        attrs: {
          type: 'checkbox',
          'aria-label': this._ariaLabel,
          'aria-labelledby': this._ariaLabelledBy,
          id: this.id,
          name: this.name,
          value: this.value,
          'aria-invalid': this.isInvalid,
          defaultChecked: this.defaultIsChecked,
          checked: this.isChecked,
          disabled: this.isDisabled
        },
        nativeOn: {
          change: e => this.$emit('change', !this.isChecked, e)
        }
      }),
      h(CControlBox, {
        props: this.styleProps
      }, [
        h(CBox, {
          props: {
            bg: 'white',
            transition: 'transform 250ms',
            rounded: 'full',
            h: this._height,
            w: this._height
          }
        })
      ])
    ])
  }
}

export default CSwitch
