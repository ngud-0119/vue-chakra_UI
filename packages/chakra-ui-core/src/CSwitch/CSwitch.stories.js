import { action } from '@storybook/addon-actions'
import { storiesOf } from '@storybook/vue'
import { CBox, CSwitch } from '..'

storiesOf('UI | Switch', module)
  .add('Basic Usage', () => ({
    components: { CBox, CSwitch },
    template: `
      <CBox mb="3">
        <CSwitch
          size="sm"
          @change="action"
          color="green"
          mr="3"
        />
        <CSwitch
          size="md"
          @change="action"
          color="blue"
          mr="3"
        />
        <CSwitch
          size="lg"
          @change="action"
          color="cyan"
        />
      </CBox>
    `,
    methods: { action: action('@change(event)') }
  }))
