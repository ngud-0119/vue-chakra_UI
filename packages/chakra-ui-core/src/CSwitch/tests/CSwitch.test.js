import CSwitch from '..'
import { render, userEvent } from '@/tests/test-utils'

const renderComponent = (props) => {
  const base = {
    components: { CSwitch },
    ...props
  }
  return render(base)
}
it('should render correctly', () => {
  const { asFragment } = renderComponent(
    {
      template: `
    <CSwitch />
    `
    }
  )
  expect(asFragment()).toMatchSnapshot()
})

it('should switch', async () => {
  const { getByTestId, getByRole } = renderComponent(
    {
      template: '<CSwitch data-testid="label" />'
    }
  )
  const label = getByTestId('label')
  const input = getByRole('checkbox')
  await userEvent.click(label)
  expect(input).toBeChecked()
})

it('should emit a change event', async () => {
  const spy = jest.fn()
  const { getByTestId } = renderComponent(
    {
      methods: {
        handleChange: spy
      },
      template: '<CSwitch data-testid="label" @change="handleChange"  /> '
    }
  )

  const label = getByTestId('label')
  await userEvent.click(label)

  expect(spy).toHaveBeenCalledTimes(1)
})

test('properly handles v-model', async () => {
  const { getByTestId, getByText } = renderComponent(
    {
      data: () => ({
        enable: false
      }),
      template: `
      <div>
        <span>{{enable ? 'enabled' : 'disabled'}}</span>
        <CSwitch v-model="enable" data-testid="label" />
      </div>`
    }
  )
  const label = getByTestId('label')

  expect(getByText('disabled')).toBeInTheDocument()

  await userEvent.click(label)

  expect(getByText('enabled')).toBeInTheDocument()
})
