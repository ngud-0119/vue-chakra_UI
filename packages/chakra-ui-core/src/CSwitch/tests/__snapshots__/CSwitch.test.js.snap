// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render correctly 1`] = `
<DocumentFragment>
  .emotion-0 {
  position: relative;
  display: inline-block;
  vertical-align: middle;
}

.emotion-1 {
  border: 0px;
  clip: rect(0px, 0px, 0px, 0px);
  height: 1px;
  width: 1px;
  margin: -1px;
  padding: 0px;
  overflow: hidden;
  white-space: nowrap;
  position: absolute;
}

.emotion-2 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: start;
  -ms-flex-pack: start;
  -webkit-justify-content: flex-start;
  justify-content: flex-start;
  -webkit-transition: all 120ms;
  transition: all 120ms;
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
  width: 1.875rem;
  height: 1rem;
  border-radius: var(--chakra-radii-full);
  background: var(--chakra-colors-gray-300);
  box-sizing: content-box;
  padding: 2px;
}

.emotion-2[aria-checked=true],
.emotion-2[data-checked] {
  background: var(--chakra-colors-blue-500);
}

.emotion-2:focus,
.emotion-2[data-focus] {
  box-shadow: var(--chakra-shadows-outline);
}

.emotion-2:hover,
.emotion-2[data-hover] {
  cursor: pointer;
}

.emotion-2[disabled],
.emotion-2[aria-disabled=true],
.emotion-2[data-disabled] {
  opacity: 0.4;
  cursor: not-allowed;
}

input[type=checkbox]:focus+.emotion-3 {
  box-shadow: var(--chakra-shadows-outline);
}

input[type=checkbox]:hover:not(:disabled):not(:checked)+.emotion-3 {
  cursor: pointer;
}

input[type=checkbox]:disabled+.emotion-3 {
  opacity: 0.4;
  cursor: not-allowed;
}

input[type=checkbox]:checked:hover:not(:disabled)+.emotion-3,
input[type=checkbox][aria-checked=mixed]:hover:not(:disabled)+.emotion-3 {
  cursor: pointer;
}

.emotion-3>* {
  -webkit-transform: translateX(0);
  -moz-transform: translateX(0);
  -ms-transform: translateX(0);
  transform: translateX(0);
}

input[type=checkbox]:checked+.emotion-3,
input[type=checkbox][aria-checked=mixed]+.emotion-3 {
  background: var(--chakra-colors-blue-500);
}

input[type=checkbox]:checked+.emotion-3>*,
input[type=checkbox][aria-checked=mixed]+.emotion-3>* {
  -webkit-transform: translateX(calc(1.875rem - 1rem));
  -moz-transform: translateX(calc(1.875rem - 1rem));
  -ms-transform: translateX(calc(1.875rem - 1rem));
  transform: translateX(calc(1.875rem - 1rem));
}

.emotion-4 {
  background: var(--chakra-colors-white);
  -webkit-transition: -webkit-transform 250ms;
  transition: transform 250ms;
  border-radius: var(--chakra-radii-full);
  height: 1rem;
  width: 1rem;
}

<label
    class="emotion-0"
    data-chakra-component="CSwitch"
    data-testid="switch"
  >
    <input
      class="emotion-1"
      data-chakra-component="CVisuallyHidden"
      type="checkbox"
    />
    <div
      aria-hidden="true"
      class="emotion-2 emotion-3"
      data-chakra-component="CControlBox"
    >
      <div
        class="emotion-4"
        data-chakra-component="CBox"
      />
    </div>
  </label>
</DocumentFragment>
`;
